src/lib/datatypes/include/sol-memdesc.h

```diff
@@ -0,0 +1,1010 @@
+     * can be used to provide defaults.
+     *
+     * Note that complex types SOL_MEMDESC_TYPE_STRUCTURE and
+     * SOL_MEMDESC_TYPE_ARRAY have their own handling with struct
+     * sol_memdesc::children.
+     */
+    union {
+        uint8_t u8; /**< @brief use when SOL_MEMDESC_TYPE_UINT8 */
+        uint16_t u16; /**< @brief use when SOL_MEMDESC_TYPE_UINT16 */
+        uint32_t u32; /**< @brief use when SOL_MEMDESC_TYPE_UINT32 */
+        uint64_t u64; /**< @brief use when SOL_MEMDESC_TYPE_UINT64 */
+        unsigned long ul; /**< @brief use when SOL_MEMDESC_TYPE_ULONG */
+        size_t sz; /**< @brief use when SOL_MEMDESC_TYPE_SIZE */
+        int8_t i8; /**< @brief use when SOL_MEMDESC_TYPE_INT8 */
+        int16_t i16; /**< @brief use when SOL_MEMDESC_TYPE_INT16 */
+        int32_t i32; /**< @brief use when SOL_MEMDESC_TYPE_INT32 */
+        int64_t i64; /**< @brief use when SOL_MEMDESC_TYPE_INT64 */
+        long l; /**< @brief use when SOL_MEMDESC_TYPE_LONG */
+        ssize_t ssz; /**< @brief use when SOL_MEMDESC_TYPE_SSIZE */
+        bool b; /**< @brief use when SOL_MEMDESC_TYPE_BOOLEAN */
+        double d; /**< @brief use when SOL_MEMDESC_TYPE_DOUBLE */
+        const char *s; /**< @brief use when SOL_MEMDESC_TYPE_STRING or SOL_MEMDESC_TYPE_CONST_STRING */
+        const void *p; /**< @brief use when SOL_MEMDESC_TYPE_PTR, SOL_MEMDESC_TYPE_STRUCTURE or SOL_MEMDESC_TYPE_ARRAY */
+    } defcontent;
+    /**
+     * @brief how to access complex types (structures and arrays).
+     *
+     * If the memory is complex, use a recursive description
+     * specified here.
+     */
+    const struct sol_memdesc *children;
+
+    /**
+     * @brief override operations to be used in this memory description.
+     *
+     * By default the operations will be done in a fixed way unless
+     * overriden by an @c ops structure, this may be used to correlate
+     * members in a structure, such as struct sol_vector where length
+     * is a member and the contents is another, with element_size
+     * being specified in yet-another. Then things like "copy" will
+     * not be a simple copy of each member.
+     *
+     * To map struct sol_vector, use SOL_MEMDESC_OPS_VECTOR. to map
+     * struct sol_ptr_vector use SOL_MEMDESC_OPS_PTR_VECTOR.
+     */
+    const struct sol_memdesc_ops {
+#ifndef SOL_NO_API_VERSION
+#define SOL_MEMDESC_OPS_API_VERSION (1) /**< API version to use in struct sol_memdesc_ops::api_version */
+        uint16_t api_version; /**< @brief API version, must match SOL_MEMDESC_OPS_API_VERSION at runtime */
+#endif
+        /**
+         * @brief initialize the defaults of memory.
+         *
+         * If provided, will be used to initialize the memory instead of
+         * the traditional use of struct sol_memdesc::defcontent.
+         *
+         * Should return 0 on success, negative errno on errors.
+         */
+        int (*init_defaults)(const struct sol_memdesc *desc, void *container);
```

Comment:
thanks for your comment, I was so biased that I forgot to review this. When I created it I was basically looking at structures and wrote the code assuming that. However the container should only be used for arrays. And with that, some properties like "name" and "offset".

I'll redesign this, have a `struct sol_memdesc_structure_member`, it will be the one with the name, offset and few other flags that doesn't make sense for all memory descriptions. Should also help save memory